symengine/assumptions.cpp

```diff
@@ -24,10 +25,86 @@ Assumptions::Assumptions(const set_basic &statements)
+                } else if (down_cast<const Number &>(*arg1).is_zero()) {
+                    set_map(nonnegative_, arg2, true);
+                    set_map(negative_, arg2, false);
+                }
+            } else if (is_a<Symbol>(*arg1) and is_a_Number(*arg2)) {
+                real_symbols_.insert(arg1);
+                if (down_cast<const Number &>(*arg2).is_negative()) {
+                    set_map(nonnegative_, arg1, false);
+                    set_map(positive_, arg1, false);
+                    set_map(negative_, arg1, true);
+                    set_map(nonpositive_, arg1, true);
+                    set_map(nonzero_, arg1, true);
+                } else if (down_cast<const Number &>(*arg2).is_zero()) {
+                    set_map(nonpositive_, arg1, true);
+                    set_map(positive_, arg1, false);
+                }
+            }
+        } else if (is_a<StrictLessThan>(*s)) {
+            const StrictLessThan &strictly_less_than
+                = down_cast<const StrictLessThan &>(*s);
+            const auto arg1 = strictly_less_than.get_arg1();
+            const auto arg2 = strictly_less_than.get_arg2();
+            if (is_a<Symbol>(*arg2) and is_a_Number(*arg1)) {
+                real_symbols_.insert(arg2);
+                if (not down_cast<const Number &>(*arg1).is_negative()) {
+                    set_map(nonnegative_, arg2, true);
+                    set_map(positive_, arg2, true);
+                    set_map(negative_, arg2, false);
+                    set_map(nonpositive_, arg2, false);
+                    set_map(nonzero_, arg2, true);
+                }
+            } else if (is_a<Symbol>(*arg1) and is_a_Number(*arg2)) {
+                real_symbols_.insert(arg1);
+                if (not down_cast<const Number &>(*arg2).is_positive()) {
+                    set_map(nonnegative_, arg1, false);
+                    set_map(positive_, arg1, false);
+                    set_map(negative_, arg1, true);
+                    set_map(nonpositive_, arg1, true);
+                    set_map(nonzero_, arg1, true);
+                }
+            }
         }
     }
 }
 
+void Assumptions::set_map(umap_basic_bool &map, const RCP<const Basic> &symbol,
+                          bool value)
+{
+    // Set element in map to true or false. Check for consistency within map
+    tribool old_value = from_map(map, symbol);
+    if ((is_true(old_value) and not value) or (is_false(old_value) and value)) {
+        throw SymEngineException("Symbol " + symbol->__str__()
+                                 + " have inconsistent positive/negativeness");
+    }
+    map[symbol] = value;
+}
+
+tribool Assumptions::from_map(const umap_basic_bool &map,
+                              const RCP<const Basic> &symbol) const
+{
+    try {
+        bool b = map.at(symbol);
+        return (tribool)b;
+    } catch (std::out_of_range const &e) {
+        return tribool::indeterminate;
+    }
```

Comment:
I like to avoid exceptions like this. My suggestion is to do,
```suggestion
    auto it = map.find(symbol);
    if (it != map.end()) {
        return (tribool)(*it);
    } else {
        return tribool::indeterminate;
    }
```

Reply:
Ok, sounds good.